Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData

sysrqb opened the pull request:
#### 714cbd04ecd4d8a57932d8e8d4f6af3a4f81dfe3
<pre>
Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData
<a href="https://bugs.webkit.org/show_bug.cgi?id=251048">https://bugs.webkit.org/show_bug.cgi?id=251048</a>
rdar://104578586

Reviewed by Chris Dumez.

This change should not make any noticable difference, but it will be needed in
later patches. In that use case, SecurityOriginData will be used in a hashmap
and that hashmap must correctly take the opaqueOriginIdentifier into account.

* Source/WTF/wtf/Markable.h:
(WTF::add):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::SecurityOrigin):
(WebCore::SecurityOrigin::isSameOriginDomain const):
(WebCore::SecurityOrigin::isSameOriginAs const):
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::equal const):
* Source/WebCore/page/SecurityOrigin.h:
(WebCore::SecurityOrigin::isOpaque const):
(WebCore::add):
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::securityOrigin const):
(WebCore::SecurityOriginData::isolatedCopy const):
(WebCore::SecurityOriginData::isolatedCopy):
(WebCore::operator==):
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::SecurityOriginData):
(WebCore::SecurityOriginData::createOpaque):
(WebCore::SecurityOriginData::isOpaque const):
(WebCore::SecurityOriginData::encode const):
(WebCore::SecurityOriginData::decode):
(WebCore::add):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/259432@main">https://commits.webkit.org/259432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/731122b8b748f0c8626456b94709ffd7dbaaf08f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37714 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114086 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174286 "Failed to checkout and rebase branch from PR 9000") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108724 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4821 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97143 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113007 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11615 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94620 "Found 1 new API test failure: TestWebKitAPI.ProcessSwap.GetUserMediaCaptureState (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39137 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93475 "Found 1 new API test failure: TestWebKitAPI.ProcessSwap.GetUserMediaCaptureState (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26234 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80801 "Found 4 new API test failures: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/usermedia-enumeratedevices-permission-check, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/audio-usermedia-permission-request, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/usermedia-permission-requests, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/display-usermedia-permission-request (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94710 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7243 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27594 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92706 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4976 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7342 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4191 "Found 1 new test failure: fast/forms/fieldset/fieldset-flexbox.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30228 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47144 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101397 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/6509 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9128 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25302 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3454 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -76,11 +89,11 @@ struct SecurityOriginData {
     bool isOpaque() const
     {
-        return protocol == emptyString() && host == emptyString() && !port;
+        return protocol == emptyString() && host == emptyString() && !port && !!opaqueOriginIdentifier;
```
Ideally, this would be simply:
`return !!opaqueOriginIdentifier;`

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -34,11 +36,16 @@ class Frame;
+
     SecurityOriginData() = default;
-    SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port)
+    SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port, MarkableOpaqueOriginIdentifier opaqueOriginIdentifier = { })
```
I think it would be less confusing to add a separate constructor that takes only an opaqueOriginIdentifier since the other parameters don't make sense when the origin is unique.

sysrqb replied:
Good point. We'll need something like this for the IPC decoder, but we can make this constructor private and leave the original constructor as public.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -76,11 +89,11 @@ struct SecurityOriginData {
     bool isNull() const
     {
-        return protocol.isNull() && host.isNull() && port == std::nullopt;
+        return protocol.isNull() && host.isNull() && port == std::nullopt && !opaqueOriginIdentifier;
```
We shouldn't need this change since protocol/host wouldn't be null if opaqueOriginIdentifier was non-null.

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,6 +49,14 @@ struct SecurityOriginData {
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
     }
+    SecurityOriginData(MarkableOpaqueOriginIdentifier opaqueOriginIdentifier)
```
Seems like this should take an `OpaqueOriginIdentifier` and be marked as explicit?

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -95,14 +116,24 @@ struct SecurityOriginData {
 #endif
+private:
+    friend struct IPC::ArgumentCoder<SecurityOriginData, void>;
+    SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port, MarkableOpaqueOriginIdentifier opaqueOriginIdentifier)
```
This constructor is overly dangerous and should probably not exist. When IPC comes from the WebProcess, it cannot be trusted, this means that IPC from a compromised WebProcess could construct a SecurityOriginData object that is in a weird state (and we could lead to bad behavior or crashes). By weird state, I mean for example that opaqueOriginIdentifier could be non-empty and yet we could have a valid-looking protocol / host.

My personal preference would be to drop this constructor and write our own encode / decode functions that would look like so:
```
encode()
{
    encoder << opaqueOriginIdentifier;
    if (opaqueOriginIdentifier)
        return;
   encoder << protocol << host << port;
}

decode()
{
    std::optional<MarkableOpaqueOriginIdentifier> opaqueOriginIdentifier;
    decoder >> opaqueOriginIdentifier;
    if (!opaqueOriginIdentifier)
        return std::nullopt;
    if (*opaqueOriginIdentifier)
        return SecurityOriginData(**opaqueOriginIdentifier);
    // decode protocol, host and port
    if (!protocol->isHashTableDeletedValue())
        return std::nullopt;
    return SecurityOriginData(protocol, host, port);
}
```

sysrqb replied:
Good point, that's a much better idea. Thanks

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -42,6 +49,14 @@ struct SecurityOriginData {
+        , port(std::nullopt)
+        , opaqueOriginIdentifier(opaqueOriginIdentifier)
+    {
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
```
Why this new RELEASE assertion? Can we omit or turn it into a debug assertion?

sysrqb replied:
It shouldn't be needed, given that we're explicitly initializing the value. I'll delete it.

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -650,6 +650,39 @@ bool ArgumentCoder<ResourceError>::decode(Decoder& decoder, ResourceError& resou
+    decoder >> port;
+    if (!port)
+        return std::nullopt;
+    return SecurityOriginData(*protocol, *host, *port);
```
`WTFMove(*protocol), WTFMove(*port)`

cdumez commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -596,13 +598,17 @@ Ref<SecurityOrigin> SecurityOrigin::create(const String& protocol, const String&
 }
 
-Ref<SecurityOrigin> SecurityOrigin::create(WebCore::SecurityOriginData&& data, String&& domain, String&& filePath, Markable<OpaqueOriginIdentifier, OpaqueOriginIdentifier::MarkableTraits>&& opaqueOriginIdentifier, bool universalAccess, bool domainWasSetInDOM, bool canLoadLocalResources, bool enforcesFilePathSeparation, bool needsStorageAccessFromFileURLsQuirk, std::optional<bool> isPotentiallyTru ...
+Ref<SecurityOrigin> SecurityOrigin::create(const SecurityOriginData& data)
```
Would we be able to use `SecurityOriginData&&` and `WTFMove(data)` below? I found one call site (not sure if there are more) and this call site seems to pass an isolated copy so it would benefit from using a r-value reference.

cdumez commented on Source/WebCore/page/SecurityOrigin.h:
```diff
@@ -226,6 +222,7 @@ class SecurityOrigin : public ThreadSafeRefCounted<SecurityOrigin> {
     WEBCORE_EXPORT SecurityOrigin();
     explicit SecurityOrigin(const URL&);
     explicit SecurityOrigin(const SecurityOrigin*);
+    explicit SecurityOrigin(const SecurityOriginData&);
```
Could we use `SecurityOriginData&&`?

cdumez commented on Source/WebCore/page/SecurityOriginData.cpp:
```diff
@@ -68,7 +68,7 @@ SecurityOriginData SecurityOriginData::fromFrame(Frame* frame)
 Ref<SecurityOrigin> SecurityOriginData::securityOrigin() const
 {
-    return SecurityOrigin::create(protocol.isolatedCopy(), host.isolatedCopy(), port);
+    return SecurityOrigin::create(isolatedCopy());
```
If `SecurityOrigin::create()` took a `SecurityOriginData&&`, we'd avoid some churn here.

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -650,6 +651,39 @@ bool ArgumentCoder<ResourceError>::decode(Decoder& decoder, ResourceError& resou
     return true;
 }
 
+void ArgumentCoder<SecurityOriginData>::encode(Encoder& encoder, const SecurityOriginData& data)
```
Ideally, these would be data members on SecurityOriginData so that we wouldn't have to include the large WebCoreArgumentCoders.h in some places. You can use ObjectIdentifier.h as an example.

cdumez commented on Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "Logging.h"
 #include "WebAuthenticationFlags.h"
 #include "WebAuthenticatorCoordinatorProxyMessages.h"
+#include "WebCoreArgumentCoders.h"
```
Should no longer be needed if you move the coders to SecurityOriginData.h.

cdumez reviewed (approved):
LGTM with nits.

cdumez commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -176,7 +176,14 @@ bool shouldTreatAsPotentiallyTrustworthy(const URL& url)
+        m_filePath = url.fileSystemPath(); // In case enforceFilePathSeparation() is called.
+}
+
+SecurityOrigin::SecurityOrigin(const SecurityOriginData& data)
```
Why not `SecurityOriginData&&` and `WTFMove(data)` below? Is there a call site that does't pass a r-value reference?

cdumez commented on Source/WebCore/page/SecurityOriginData.h:
```diff
@@ -95,14 +115,47 @@ struct SecurityOriginData {
     String debugString() const { return toString(); }
 #endif
+
+    template<class Encoder> void encode(Encoder& encoder) const
```
Those are really large, it would be better to move them below the class instead of putting them inline like this.

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.h:
```diff
@@ -161,6 +161,7 @@ struct PasteboardImage;
 struct PromisedAttachmentInfo;
 struct RecentSearch;
 struct ScrollableAreaParameters;
+struct SecurityOriginData;
```
Is this really needed?

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
```diff
@@ -106,6 +106,7 @@
 #include <WebCore/SearchFieldPart.h>
 #include <WebCore/SearchPopupMenu.h>
 #include <WebCore/SecurityOrigin.h>
+#include <WebCore/SecurityOriginData.h>
```
Is this really needed?

sysrqb committed:
Move Opaque Origin Identifier from SecurityOrigin to SecurityOriginData
https://bugs.webkit.org/show_bug.cgi?id=251048
rdar://104578586

Reviewed by Chris Dumez.

This change should not make any noticable difference, but it will be needed in
later patches. In that use case, SecurityOriginData will be used in a hashmap
and that hashmap must correctly take the opaqueOriginIdentifier into account.

* Source/WTF/wtf/Markable.h:
(WTF::add):
* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::SecurityOrigin::SecurityOrigin):
(WebCore::SecurityOrigin::isSameOriginDomain const):
(WebCore::SecurityOrigin::isSameOriginAs const):
(WebCore::SecurityOrigin::create):
(WebCore::SecurityOrigin::equal const):
* Source/WebCore/page/SecurityOrigin.h:
(WebCore::SecurityOrigin::isOpaque const):
(WebCore::add):
* Source/WebCore/page/SecurityOriginData.cpp:
(WebCore::SecurityOriginData::securityOrigin const):
(WebCore::SecurityOriginData::isolatedCopy const):
(WebCore::SecurityOriginData::isolatedCopy):
(WebCore::operator==):
* Source/WebCore/page/SecurityOriginData.h:
(WebCore::SecurityOriginData::SecurityOriginData):
(WebCore::SecurityOriginData::createOpaque):
(WebCore::SecurityOriginData::isOpaque const):
(WebCore::SecurityOriginData::encode const):
(WebCore::SecurityOriginData::decode):
(WebCore::add):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/259432@main
